File: solutions/leetcode/product-of-array-except-self.cpp
/* Problem product-of-array-except-self
 * https://www.leetcode.com/problems/product-of-array-except-self
 */

#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> productExceptSelf(vector<int>& nums) {
    size_t const n = nums.size();
    vector<int> pre(n + 1), su(n + 1);
    pre[0] = 1, su[n] = 1;
    for (int i = 0; i < n; ++i) {
      pre[i + 1] = pre[i] * nums[i];
      su[n - i - 1] = su[n - i] * nums[n - i - 1];
    }
    for (int i = 0; i < n; ++i) {
      nums[i] = pre[i] * su[i + 1];
    }
    return nums;
  }
};

File: solutions/leetcode/product-of-array-except-self.rs
/* Problem product-of-array-except-self
 * https://www.leetcode.com/problems/product-of-array-except-self
 */

struct Solution;

impl Solution {
	pub fn product_except_self(mut nums: Vec<i32>) -> Vec<i32> {
		let (exist_zero, product) =
			nums.iter()
				.fold((false, 1i32), |(exist_zero, product), num| {
					if *num == 0 && !exist_zero {
						(true, product)
					} else {
						(exist_zero, product * num)
					}
				});

		for num in &mut nums {
			*num = if *num == 0 {
				product
			} else if exist_zero {
				0
			} else {
				product / *num
			};
		}

		nums
	}
}

pub fn main() {
	assert_eq!(
		vec![24, 12, 8, 6],
		Solution::product_except_self(vec![1, 2, 3, 4])
	);
	assert_eq!(
		vec![0, 0, 9, 0, 0],
		Solution::product_except_self(vec![-1, 1, 0, -3, 3])
	);
}
